Systematic handling of remotely required tool files.

jmchilton opened the pull request:
- Allow tools to define "pulsar file action"-like paths for files to include and exclude during remote job setup.
- Package up a list of defaults for older tools to allow older tool versions to work with Pulsar without hacking the tool XML.

Fixes https://github.com/galaxyproject/pulsar/issues/113 and alternative to https://github.com/galaxyproject/pulsar/pull/260. 

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

mvdbeek commented on test/unit/tool_util/test_required_files.py:
```diff
@@ -0,0 +1,170 @@
+TOOL_REQUIRED_FILES_XML_1 = """
+<tool name="required_files_1" id="required_files_1" version="0.1.0">
+    <command>foo</command>
+    <required_files>
```
I really like this, and i think we can use this in planemo as well for containerized jobs that need files that are symlinked in. Not sure how exactly that would look like but this would make it easier to find a solution that is not what we did now ... copying everything.

jmchilton committed:
Systematic handling of remotely required tool files.

- Allow tools to define "pulsar file action"-like paths for files to include and exclude during remote job setup.
- Package up a list of defaults for older tools.

natefoo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -191,6 +192,16 @@
     "winSplitter": packaging.version.parse("1.0.1"),
 }
 
+REQUIRE_FULL_DIRECTORY = {
```
Handling this doesn't appear to be implemented?

jmchilton replied:
I think it is - see comment below?

jmchilton replied:
Have you tried it and it doesn’t work?

natefoo replied:
Nevermind, I see how it works now, thanks.

natefoo reviewed (approved):
Rebased and ready to roll.
